global: circulation exceptions

topless opened the pull request:
* overridable exception name
* overridable exception description
* global logging
* exceptions: application exceptions
* exceptions: rest exceptions
* exceptions: fixed tests
* addresses inveniosoftware/invenio-app-ils#193

zzacharo commented on invenio_circulation/errors.py:
```diff
@@ -8,50 +8,156 @@
-        """Constructor."""
-        self.msg = msg
 
+class CirculationException(RESTValidationError):
```
What we gain by inheriting `RESTValidationError`? I mean this used [here](https://github.com/inveniosoftware/invenio-records-rest/blob/ada7d4d556d7b92cab88039bbfa87e4781e320ab/invenio_records_rest/facets.py#L47) really for displaying the marshmallow errors. Wouldn't https://github.com/inveniosoftware/invenio-rest/blob/4271708f0e2877e5100236be9242035b95b5ae6e/invenio_rest/errors.py#L46 be enough?

zzacharo commented on invenio_circulation/errors.py:
```diff
@@ -8,50 +8,156 @@
+    """Circulation error codes."""
+
+    # Permissions
+    InvalidPermission = 10
```
Wouldn't be nicer to write all members of the enum with capitals and underscores in between?

zzacharo commented on invenio_circulation/ext.py:
```diff
@@ -118,10 +118,9 @@ def trigger(self, loan, **kwargs):
                 return loan
             except TransitionConditionsFailed as ex:
-                current_app.logger.debug(ex.msg)
+                # NOTE: Silent pass not good, gets logged by Exception itself.
```
How we can omit that?

zzacharo commented on invenio_circulation/transitions/base.py:
```diff
@@ -30,22 +29,22 @@ def inner(self, loan, **kwargs):
-            raise TransitionConstraintsViolation(msg=msg)
+            msg = "Loan item is '{0}' but transition is trying to set it to " \
+                  "'{1}'".format(loan['item_pid'], new_item_pid)
+            raise TransitionConstraintsViolation(description=msg)
```
I think this should be set to a new exception as by raising `TransitionConstraintsViolation` it will trigger the check of the next available transition and probably we need to exit here? WDYT @ntarocco @kprzerwa ?

zzacharo commented on invenio_circulation/transitions/transitions.py:
```diff
@@ -70,8 +69,8 @@ def _ensure_valid_extension(loan):
-        msg = 'Max extension count reached `{0}`'.format(extension_max_count)
-        raise TransitionConstraintsViolation(msg=msg)
+        msg = "Max extension count reached '{0}'".format(extension_max_count)
+        raise TransitionConstraintsViolation(description=msg)
```
Also here?

topless replied:
I have created different Exceptions for these as suggested.

zzacharo commented on invenio_circulation/views.py:
```diff
@@ -8,43 +8,22 @@
-HTTP_CODES = {"bad_request": 400, "accepted": 202}
-
-
-def create_error_handlers(blueprint):
```
how the rest exceptions are handled if you delete this?

ntarocco commented on invenio_circulation/errors.py:
```diff
@@ -8,50 +8,142 @@
-        """Constructor."""
-        self.msg = msg
 
+class CirculationException(RESTException):
```
Just cosmetic: I would align with Invenio naming, exceptions in general have the suffix `Error`, plus it is aligned with [PEP8](https://www.python.org/dev/peps/pep-0008/#exception-names) (I have also double checked with Alex). See here: https://github.com/inveniosoftware/invenio-pidstore/blob/master/invenio_pidstore/errors.py
Can we quickly changed all of them?

ntarocco commented on invenio_circulation/errors.py:
```diff
@@ -8,50 +8,142 @@
 
-class ItemNotAvailable(CirculationException):
-    """Exception raised from action on unavailable item."""
+    def __init__(self, **kwargs):
```
Again cosmetic: put constructor as first method?

ntarocco commented on invenio_circulation/errors.py:
```diff
@@ -8,50 +8,142 @@
+        data["error_module"] = "Circulation"
+        data["error_class"] = self.name
+        resp.data = json.dumps(data)
+        current_app.logger.exception(self)
```
I think we should not log it, because otherwise we will always have a Sentry exception logged.
For unhandled exception, it should happen, but not for handled...

ntarocco commented on invenio_circulation/errors.py:
```diff
@@ -8,50 +8,142 @@
+        """Initialize exception."""
+        super(CirculationException, self).__init__(**kwargs)
 
+    def get_response(self, environ=None, **kwargs):
```
I think it is better to override `get_body` rather than `get_response`, no?
See it here: https://github.com/inveniosoftware/invenio-rest/blob/master/invenio_rest/errors.py#L68

Example pseudo-code:
```
def get_body(...):
    string_body = super.get_body()
    body = json.loads(string_body)
    body["error_module"] = "Circulation"
    body["error_class"] = self.name
    return json.dumps(body)
```

ntarocco commented on invenio_circulation/ext.py:
```diff
@@ -118,10 +118,9 @@ def trigger(self, loan, **kwargs):
                 return loan
             except TransitionConditionsFailed as ex:
-                current_app.logger.debug(ex.msg)
+                current_app.logger.exception(ex)
```
👍

ntarocco commented on tests/test_errors.py:
```diff
@@ -0,0 +1,153 @@
+    assert ex.value.description == msg
+
+
+def test_loan_max_extensio(app):
```
small typo, `extensions`

ntarocco commented on invenio_circulation/permissions.py:
```diff
@@ -17,17 +17,19 @@
     if permission is not None and not permission.can():
         if current_user.is_authenticated:
-            abort(403, 'You do not have a permission for this action')
+            raise InvalidPermission(action=action, permission=permission)
```
I would remove this one, since it is a generic Invenio exception, otherwise the circulation module will behave differently for REST Api permissions checks.
I would also **NOT** expose the action name.

topless replied:
Meaning remove it completely or revert it to `abort(403)`?

kpsherva commented on invenio_circulation/errors.py:
```diff
@@ -8,50 +8,142 @@
-    """Exception raised when item can not be requested."""
+
+# Permissions
+class InvalidPermission(CirculationException):
```
it would be nice if you add `Error` suffix to all the exception classes for readability

kpsherva commented on invenio_circulation/search/api.py:
```diff
@@ -42,7 +44,10 @@ def search_by_pid(item_pid=None, document_pid=None, filter_states=None,
         search = search.filter("term", item_pid=item_pid)
     else:
-        raise ValueError("Must specify item_pid or document_pid param")
+        raise PropertyRequired(description=(
```
I think it should be named `RequiredParameterMissingError` not property. Property for me is something that an object owns, and we pass those values as search parameters.

kpsherva commented on invenio_circulation/errors.py:
```diff
@@ -8,50 +8,142 @@
+    """Exception raised when item can not be requested."""
+
+
+class NotImplementedCirculation(CirculationException):
```
I think the name is a bit misleading in comparison when this exception is raised, maybe as something like `ConfigurationVariableMissing`?

topless committed:
release: v1.0.0a9

kpsherva commented on invenio_circulation/permissions.py:
```diff
@@ -17,18 +17,20 @@
 
 
-def check_permission(permission):
+def check_permission(action, permission):
```
unused param action

kpsherva commented on invenio_circulation/transitions/base.py:
```diff
@@ -105,30 +105,29 @@ def __init__(self, src, dest, trigger='next', permission_factory=None,
         if self.permission_factory and not self.permission_factory(loan).can():
-            msg = 'Invalid circulation permission'
-            raise InvalidCirculationPermission(msg=msg)
+            raise InvalidPermissionError()
```
no description?

topless replied:
If we don't provide description it will use the default one.

zzacharo commented on invenio_circulation/transitions/base.py:
```diff
@@ -56,13 +56,13 @@ def ensure_required_params(f):
+            msg = "Required input parameters are missing '[{}]'" \
                 .format(missing)
-            raise TransitionConstraintsViolation(msg=msg)
+            raise TransitionConstraintsViolationError(description=msg)
```
I think here we should not raise `TransitionConstraintsViolationError` as this exception is used to loop over the available transitions when you trigger a new one. If you have missing a required param this will be valid for every transition. So maybe we need to raise something like `NoRequiredParamsAvailableError`. WDYT @ntarocco @kprzerwa ?

topless replied:
@zzacharo we can raise the existing exception `MissingRequiredParameterError`.

zzacharo replied:
Yes sure!

zzacharo commented on invenio_circulation/transitions/base.py:
```diff
@@ -56,13 +57,13 @@ def ensure_required_params(f):
+            msg = "One of the parameters '[{}]' must be passed."\
                 .format(missing)
-            raise TransitionConstraintsViolation(msg=msg)
+            raise TransitionConstraintsViolationError(description=msg)
```
@topless can you also change it here? I think is the same case!

topless committed:
global: circulation exceptions

* overridable exception name
* overridable exception description
* global logging
* application exceptions
* migrated existing generic exceptions to app specific
* rest exceptions
* fixed tests
* review refactor
* addresses #193


fix: typo in exception class


fix: names and args


fix: changed exception to MissingRequiredParameterError


fix: missing partial params


fix: missing params


fix: broken test
